rust/flatbuffers/src/builder.rs

```diff
@@ -235,6 +238,49 @@ impl<'fbb> FlatBufferBuilder<'fbb> {
         WIPOffset::new(o.value())
     }
 
+    #[inline]
+    pub fn create_shared_string<'a: 'b, 'b>(&'a mut self, s: &'b str) -> WIPOffset<&'fbb str> {
+        self.assert_not_nested(
+            "create_shared_string can not be called when a table or vector is under construction",
+        );
+
+        // Saves a ref to owned_buf since rust doesnt like us refrencing it
+        // in the binary_search_by code.
+        let buf = &self.owned_buf;
+
+        let found = self.strings_pool.binary_search_by(|offset| {
+            let ptr = offset.value() as usize;
+            // Gets The pointer to the size of the string
+            let str_memory = &buf[buf.len() - ptr..];
+            // Gets the size of the written string from buffer
+            let size = u32::from_le_bytes([
+                str_memory[0],
+                str_memory[1],
+                str_memory[2],
+                str_memory[3],
+            ]) as usize;
+            // Size of the string size
+            let string_size: usize = 4;
+            // Fetches actual string bytes from index of string after string size
+            // to the size of string plus string size
+            let iter = str_memory[string_size..size + string_size].iter();
+            // Compares bytes of fetched string and current writable string
+            iter.cloned().cmp(s.bytes())
+        });
+
+        let address = if let Ok(index) = found {
+            self.strings_pool[index]
+        } else {
+            let _address = WIPOffset::new(self.create_byte_string(s.as_bytes()).value());
+            _address
+        };
+
+        if let Err(index) = found {
```

Comment:
This `if let` block is executed iff the above `else` is executed. The blocks should be merged with a `match` expression. Also note that the last expression of the function is returned so the `return` keyword isn't needed
```rust
match found {
  Ok(index) => self.strings_pool[index],
  Err(index) => {
    let address = WIPOffset::new(self.create_byte_string(s.as_bytes()).value());
    self.strings_pool.insert(index, address);
    address
  }
}
 ```